src/NewRemoting/MessageHandler.cs

```diff
@@ -41,10 +59,215 @@ public MessageHandler(InstanceManager instanceManager, FormatterFactory formatte
+						bool hasReturnValue;
+						if (del.Method.ReturnType != typeof(void))
+						{
+							hasReturnValue = true;
+							arguments.Add((Type)del.Method.ReturnType);
+							switch (arguments.Count)
+							{
+								case 1:
+									proxyType = typeof(DelegateFuncProxyOnClient<>)
+										.MakeGenericType(arguments.ToArray());
+									break;
+								case 2:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 3:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 4:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 5:
+									proxyType =
+										typeof(DelegateFuncProxyOnClient<,,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								default:
+									throw new NotSupportedException(
+										$"Unsupported number of arguments for function ({arguments.Count}");
+							}
+						}
+						else
+						{
+							hasReturnValue = false;
+							switch (arguments.Count)
+							{
+								case 0:
+									proxyType = typeof(DelegateProxyOnClient);
+									break;
+								case 1:
+									proxyType = typeof(DelegateProxyOnClient<>).MakeGenericType(arguments.ToArray());
+									break;
+								case 2:
+									proxyType = typeof(DelegateProxyOnClient<,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 3:
+									proxyType = typeof(DelegateProxyOnClient<,,>).MakeGenericType(arguments.ToArray());
+									break;
+								case 4:
+									proxyType = typeof(DelegateProxyOnClient<,,,>).MakeGenericType(arguments.ToArray());
+									break;
+								default:
+									throw new NotSupportedException(
+										$"Unsupported number of arguments for action ({arguments.Count}");
+							}
+						}
+
+						var proxy = Activator.CreateInstance(proxyType);
+						var addEventMethod = proxyType.GetMethod("add_Event");
+						addEventMethod.Invoke(proxy, new[] { del });
+						_instanceManager.AddInstance(proxy, instanceId, referencesWillBeSentTo, proxyType);
+
+						w.Write((int)RemotingReferenceType.AddEvent);
+						LogMsg(RemotingReferenceType.AddEvent);
+						w.Write(instanceId);
+						w.Write(hasReturnValue);
+						// If the type of the delegate method is generic, we need to provide its type arguments
+						w.Write(arguments.Count);
+						foreach (var argType in arguments)
+						{
+							string arg = argType.AssemblyQualifiedName;
+							if (arg == null)
+							{
+								throw new RemotingException("Unresolved generic type or some other undefined case");
+							}
+
+							w.Write(arg);
+						}
+					}
+				}
+				else
+				{
+					// The delegate target is a static method
+					////						w.Write(string.Empty);
+					throw new InvalidOperationException("The delegate target is a static method");
```

Comment:
We should clarify the exceptions to throw here. We have a mix of `RemotingException`, `InvalidOperationException` and `NotSupportedException` for very similar errors.

Reply:
Clients should only catch RemotingExceptions. Unless it is a programming error we should only use this.

Reply:
I agree, but then please review the exceptions in this method for this. (e.g. the use of a static method would be a programming error everywhere)